README.md

```diff
@@ -28,4 +28,294 @@ One of the very common use cases of WebXR is 180 and/or 360 photos or videos, bo
+  attribute float width = 1;
+  attribute float height = 1;
+}
+
+interface XRCylinderLayer extends XRLayer {
+  attribute XRRigidTransform transform;
+  attribute float radius = 1;
+  attribute float centralAngle = Math.PI * 0.5;
+  attribute float aspectRatio = 1;
+}
+
+interface XREquirectLayer extends XRLayer {
+  attribute XRRigidTransform transform;
+  attribute float radius = 1;
+  attribute float scaleX = 1;
+  attribute float scaleY = 1;
+  attribute float biasX = 0;
+  attribute float biasY = 0;
+}
+
+interface XRCubeLayer extends XRLayer {
+  attribute DOMPoint orientation;
+}
+
+typedef (XRQuadLayer or XRCylinderLayer or XREquirectLayer or XRCubeLayer)
+        XRNonProjectionLayer; // Would love a better term for this!
+
+//
+// Graphics Bindings
+//
+
+dictionary XRWebGLLayerSourceInit {
+  boolean stereo = false;
+  boolean depth = true;
+  boolean stencil = false;
+  boolean alpha = true;
+  boolean ignoreDepthValues = false;
+}
+
+interface XRWebGLGraphicsBinding {
+  constructor(XRSession session, WebGLRenderingContext context);
+
+  double getNativeProjectionScaleFactor();
+
+  Promise<XRWebGLFramebufferLayerSource> setFramebufferLayerSource(
+      XRProjectionLayer layer, XRWebGLLayerInit init);
+  Promise<XRWebGLFramebufferLayerSource> setFramebufferLayerSource(
+      XRNonProjectionLayer layer, int pixelWidth, int pixelHeight, XRWebGLLayerSourceInit init);
+}
+
+interface XRWebGL2GraphicsBinding extends XRWebGLGraphicsBinding {
+  constructor(XRSession session, WebGL2RenderingContext context);
+
+  Promise<XRWebGLTextureLayerSource> setTextureLayerSource(
+      XRProjectionLayer layer, XRWebGLLayerInit init);
+  Promise<XRWebGLTextureLayerSource> setTextureLayerSource(
+      XRNonProjectionLayer layer, int pixelWidth, int pixelHeight, XRWebGLLayerSourceInit init);
+}
+
+//
+// Layer Sources
+//
+
+interface XRLayerSource {
+  readonly attribute unsigned long width;
+  readonly attribute unsigned long height;
+  readonly attribute unsigned long arraySize;
+
+  readonly attribute boolean ignoreDepthValues;
+}
+
+interface XRWebGLFramebufferLayerSource extends XRLayerSource {
+  // Getter function serves as flag that layer will be written to this frame.
+  WebGLFramebuffer? getCurrentFramebuffer();
+
+  readonly attribute boolean antialias;
+}
+
+interface XRWebGLTextureLayerSource extends XRLayerSource {
+  WebGLTexture? getCurrentColorTexture();
+  WebGLTexture? getCurrentDepthStencilTexture();
+}
+```
+
+## Appendix B: Open Questions
+
+  - Sensible defaults of all the layer values.
+  - Do we need an attribte to communicate max number of supported layers?
+  - Do we need to report on the layer source whether or not it's mono/stereo, since the XR device may force a downgrade?
+  - Depth: Should we consider allowing a separate layer for this like OpenXR does, or continue pattern of allowing it's use by default (implies most layer types need to be able to allocate both a color and depth/stencil buffer)
```

Comment:
Depth wise - I think, we can include the depth into the layer source for simplicity, which existence can be controlled via 'depth' (or 'has_depth') attribute. By doing this, we would also be able to discard it at the end to avoid depth resolve.

Reply:
I don't see anything about stereo layout here (side-by-side, top-to-bottom, or just stereo (for texture arrays)), but it could be added later.

Reply:
Will need to add more rendering-hinting flags, such as MSAA samples, FFR, CAC.

Reply:
Will need to have a way to signal that layer's image has changed or not: thus we might avoid extra GPU work and just reuse the image from the previous frame. In fact, since all of those 'textures' and 'framebuffers' are actually swapchains, this is a must to have, since the underlying compositor must now which slice of the swapchain to use and whether it needs to be incremented or not.

Reply:
One more thing: are all possible LayerSources should be supported? What is the way to determine which ones are and which ones are not?

Reply:
We need to design the API in such a way that non-sensical combinations can't be constructed.
Otherwise we have to define in prose which ones are valid and have to throw on invalid combinations.

Reply:
I can't comment your previous example. What would the view/projection matrices be for non-projection layers? Are there frameworks that can render stereo in equirect and cube layers?